examples/server/server.cpp

```diff
@@ -1321,17 +1321,24 @@ struct server_slot {
             && are_lora_equal(lora, other_slot.lora);
     }
 
+    //  There are two caps on the budge of a single request:
+    //  * [params.n_predict]
+    //  * [global_params.n_predict]
+    // This function returns true if the request is not limited by either of them.
     bool has_budget(const common_params & global_params) {
         if (params.n_predict == -1 && global_params.n_predict == -1) {
             return true; // limitless
         }
+        n_remaining = INT32_MAX;
 
-        n_remaining = -1;
+        // The request or server have finite limits on the number of tokens to generate.
+        if ((params.n_predict != -1 && params.n_predict != -2) || (global_params.n_predict  != -1 && global_params.n_predict != -2)) {
+            n_remaining = std::min(n_remaining, params.n_predict - n_decoded);
+        }
 
-        if (params.n_predict != -1) {
-            n_remaining = params.n_predict - n_decoded;
-        } else if (global_params.n_predict != -1) {
-            n_remaining = global_params.n_predict - n_decoded;
+        // The request or server have limits based on the context window.
+        if (params.n_predict == -2 || global_params.n_predict == -2) {
+            n_remaining = std::min(n_remaining, n_ctx - n_decoded);
```

Comment:
Make sure `n_ctx` is the slot's context (probably it's not).

Reply:
It is commented as "context size per slot" ([link](https://github.com/ggml-org/llama.cpp/pull/12323/files#diff-87355a1a297a9f0fdc86af5e2a59cae153290f58d68822cd10c30fee4f7f7076R1247)).

Is that not right?

Reply:
yes that's correct, the `slot.n_ctx` is not the global context, probably we should rename this variable in a future PR to prevent confusions

Ref. code in `server_context::init()`:

```cpp
const int32_t n_ctx_slot = n_ctx / params_base.n_parallel;
...
slot.n_ctx = n_ctx_slot;
```

Reply:
Sorry for causing the confusion - the `n_ctx` is correct here. I just had some doubts that it might be referring to the global context, but it is not the case, so all is good.